### PR TITLE
Switch to CRI v1 API

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
The container runtime containerd as well as CRI-O already support the
`v1` as well as `v1alpha2` API of the Container Runtime Interface (CRI).
This means we can do now the next step and switch to `v1` completely.

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

cc @mrunalp @mikebrow 

#### Does this PR introduce a user-facing change?

```release-note
Switch to Container Runtime Interface (CRI) API `v1` (from `v1alpha2`)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
Oh, looks like we have to wait for containerd 1.6 to be released.

**Comment by aojea** [comment]
> Oh, looks like we have to wait for containerd 1.6 to be released.

does this mean it won't work against containerd 1.5?
cc: @dims

**Comment by saschagrunert** [comment]
> > Oh, looks like we have to wait for containerd 1.6 to be released.
> 
> does this mean it won't work against containerd 1.5?
> cc: @dims

Yes, a runtime would have to support CRI v1, because the kubelet now expects that protocol to be served.

**Comment by aojea** [comment]
is the plan to not support containerd 1.5 since kubernetes 1.23?

**Comment by mikebrow** [comment]
We could cherry pick the containerd v1 CRI support to the containerd 1.5 release.

**Comment by aojea** [comment]
> We could cherry pick the containerd v1 CRI support to the containerd 1.5 release.

we are going to need some cross-coordination, jobs are pinning the containerd versions

```
 test-infra]$ grep -r CONTAINERD_VER config/
config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml:            - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml:            - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml:          - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-network/sig-network-misc.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-network/sig-network-misc.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-network/sig-network-misc.yaml:      - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.5
config/jobs/kubernetes/sig-release/release-branch-jobs/1.19.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.3.5
config/jobs/kubernetes/sig-release/release-branch-jobs/1.20.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.4.3
config/jobs/kubernetes/sig-release/release-branch-jobs/1.20.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.4.3
config/jobs/kubernetes/sig-release/release-branch-jobs/1.21.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.4.3
config/jobs/kubernetes/sig-release/release-branch-jobs/1.21.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.4.3
config/jobs/kubernetes/sig-release/release-branch-jobs/1.22.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.2
config/jobs/kubernetes/sig-release/release-branch-jobs/1.22.yaml:        - --env=KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.2
```

and a bigger announcement, since people will require to upgrade the runtime when upgrading the cluster.

**Comment by saschagrunert** [comment]
> We could cherry pick the containerd v1 CRI support to the containerd 1.5 release.

@mikebrow I'd prefer the cherry-pick to have more backwards compatible containerd versions available. 

> and a bigger announcement, since people will require to upgrade the runtime when upgrading the cluster.

@aojea sure thing, I can take care of that if the CI of this PR goes green'ish.

--- 

I also pre-evaluated ways of making the change backwards compatible. The fact that the protobuf generated types are everywhere used in the kubelet made us come to the decision that a hard-cut seems to be the best option. In theory it would be possible to "try to connect v1 and fallback to v1alpha2 client", but for that we require common types which work for both APIs (conversions [like those](https://github.com/cri-o/cri-o/blob/10db266b44488bc4f17a7af5c51303f511345bdc/server/cri/v1alpha2/rpc_create_container.go#L18)).

**Comment by mikebrow** [comment]
> > We could cherry pick the containerd v1 CRI support to the containerd 1.5 release.
> 
> @mikebrow I'd prefer the cherry-pick to have more backwards compatible containerd versions available.
> 
> > and a bigger announcement, since people will require to upgrade the runtime when upgrading the cluster.
> 
> @aojea sure thing, I can take care of that if the CI of this PR goes green'ish.
> 
> I also pre-evaluated ways of making the change backwards compatible. The fact that the protobuf generated types are everywhere used in the kubelet made us come to the decision that a hard-cut seems to be the best option. In theory it would be possible to "try to connect v1 and fallback to v1alpha2 client", but for that we require common types which work for both APIs (conversions [like those](https://github.com/cri-o/cri-o/blob/10db266b44488bc4f17a7af5c51303f511345bdc/server/cri/v1alpha2/rpc_create_container.go#L18)).

Agree, and types will become a problem when the api's diverge... Till then (or till we get go generics) we can use the v1 api and types in kubelet and switch to v1alpha at the api interface layer using the proto build marshaling https://github.com/containerd/containerd/blob/main/pkg/cri/server/instrumented_service.go#L1503-L1531. Though that won't be necessary after the supported container runtimes support both.

**Comment by BenTheElder** [comment]
Essentially all e2e jobs pin the runtime versions. We can upgrade those and KIND at least is willing to run pre-release versions, but there's definitely also the angle of operators needing to upgrade to runtime versions that aren't even out yet?

> The fact that the protobuf generated types are everywhere used in the kubelet 

That seems like a bug.

In another core k8s API I'm pretty sure using this instead of an internal type w/ conversions would be considered a bug.

I feel like this should work like k8s core APIs, with a period of multiple version support, otherwise this is going to cause operator distress of at least the order of the ""docker support is being dropped"" saga.

Alternatively kubelet could continue to use a single version but wait to do a hard cut over for some time after the runtimes start supporting both, but that seems harder to develop and validate.

**Comment by liggitt** [comment]
> The container runtime containerd as well as CRI-O already support the
> `v1` as well as `v1alpha2` API of the Container Runtime Interface (CRI).
> This means we can do now the next step and switch to `v1` completely.

Does that mean that? Are minimum versions of container runtimes clearly communicated to cluster operators? What was the overlap period where both CRI API versions were supported?

**Comment by saschagrunert** [comment]
Hm, so the alternative is to introduce intermediate types which are mapped to CRI v1 and v1alpha2. This way we can support both connections at the same time without breaking types. Later on we can diverge both versions while deprecating v1alpha2 in 1.23.

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func (m *kubeGenericRuntimeManager) generateWindowsContainerConfig(container *v1
 		// Part three - CRI & ContainerD's implementation
 		//   The kubelet sets these directly on CGroups in Linux, but needs to pass them across CRI on Windows.
 		//   There is an existing cpu_maximum field, with a range of percent * 100, so 1-10000. This is different from Docker, but consistent with OCI
-		//   https://github.com/kubernetes/kubernetes/blob/56d1c3b96d0a544130a82caad33dd57629b8a7f8/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L681-L682
+		//   https://github.com/kubernetes/kubernetes/blob/56d1c3b96d0a544130a82caad33dd57629b8a7f8/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L681-L682
```
Comment (fuweid) [inline pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:80]: maybe the blob `56d1c3b96d0a544130a82caad33dd57629b8a7f8` should be changed.

**Comment by saschagrunert** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/104575

**Pull request closed without merging**